### PR TITLE
refactor: creating mechs in marketplace

### PR DESCRIPTION
**Pull request by kupermind**
- Creating mechs in marketplace.

### CODE REVIEW

File: contracts/MechMarketplace.sol [hunk 1]
```diff
@@ -133,6 +131,21 @@ contract MechMarketplace is IErrorsMarketplace {
         );
     }
 
+    /// @dev Registers service as a mech.
+    /// @param serviceId Service id.
+    /// @param price Minimum required payment the agent accepts.
+    /// @return mech The created mech instance address.
+    function create(uint256 serviceId, uint256 price) external returns (address mech) {
+        // Get service multisig address
+        (, address multisig, , , , , ) = IServiceRegistry(serviceRegistry).mapServices(serviceId);
```
[the remark this answers is not in the document]
Comment (kupermind) [reply contracts/MechMarketplace.sol]: It will be checked later in OlasMech via the owner and multisig address

File: contracts/AgentMech.sol [hunk 2]
```diff
@@ -288,23 +288,6 @@ contract AgentMech is OlasMech {
         emit Deliver(msg.sender, requestId, requestData);
     }
 
-    /// @dev Registers a request without a marketplace.
-    /// @notice Interface provided for backwards compatibility only.
-    /// @param data Self-descriptive opaque data-blob.
-    /// @return requestId Request Id.
-    function request(bytes memory data) external payable returns (uint256 requestId) {
```
Comment (kupermind) [inline contracts/AgentMech.sol:295]: All the mechs must be based on mech marketplace now

File: contracts/AgentMech.sol [hunk 3]
```diff
@@ -342,28 +325,6 @@ contract AgentMech is OlasMech {
         emit RevokeRequest(account, requestId);
     }
 
-    /// @dev Delivers a request without a marketplace.
-    /// @notice Interface provided for backwards compatibility only.
-    /// @param requestId Request id.
-    /// @param data Self-descriptive opaque data-blob.
-    function deliver(uint256 requestId, bytes memory data) external onlyOperator {
```
Comment (kupermind) [inline contracts/AgentMech.sol:349]: All the mechs must be based on mech marketplace now

File: contracts/MechFactoryBasic.sol [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    // Agent factory version number
+    string public constant VERSION = "0.1.0";
+
+    /// @dev Registers service as a mech.
+    /// @param mechMarketplace Mech marketplace address.
+    /// @param serviceRegistry Service registry address.
+    /// @param serviceId Service id.
+    /// @param payload Mech creation payload.
+    /// @return mech The created mech instance address.
+    function createMech(
```
Comment (kupermind) [inline contracts/MechFactoryBasic.sol:19]: Basically mech creation function as before, just depends on mech passed params to decode.

File: contracts/MechMarketplace.sol [hunk 5]
```diff
@@ -21,10 +32,7 @@ struct MechDelivery {
 
 /// @title Mech Marketplace - Marketplace for posting and delivering requests served by agent mechs
 contract MechMarketplace is IErrorsMarketplace {
-    event OwnerUpdated(address indexed owner);
```
Comment (kupermind) [inline contracts/MechMarketplace.sol:24]: Outdated events

File: contracts/MechMarketplace.sol [hunk 6]
```diff
@@ -133,6 +141,16 @@ contract MechMarketplace is IErrorsMarketplace {
         );
     }
 
+    /// @dev Registers service as a mech.
+    /// @param serviceId Service id.
+    /// @param mechFactory Mech factory address.
+    /// @return mech The created mech instance address.
+    function create(uint256 serviceId, address mechFactory, bytes memory payload) external returns (address mech) {
```
Comment (kupermind) [inline contracts/MechMarketplace.sol:148]: This is so pretty now <3

File: contracts/integrations/nevermined/MechFactorySubscription.sol [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    // Agent factory version number
+    string public constant VERSION = "0.1.0";
+
+    /// @dev Registers service as a mech.
+    /// @param mechMarketplace Mech marketplace address.
+    /// @param serviceRegistry Service registry address.
+    /// @param serviceId Service id.
+    /// @param payload Mech creation payload.
+    /// @return mech The created mech instance address.
+    function createMech(
```
Comment (kupermind) [inline contracts/integrations/nevermined/MechFactorySubscription.sol:20]: Basically mech creation function as before, just depends on mech passed params to decode.

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**